### PR TITLE
[WIP] Convolutional Self-Attention Networks

### PR DESCRIPTION
**Pull request by steremma**
Inspired from the [Relative Position](https://arxiv.org/abs/1803.02155) implementation in the recent release I did some research on the attention mechanism at large and found [Convolutional Self-Attention Networks](https://arxiv.org/abs/1904.03107) interesting. While relevant advances are being published too fast to follow, the basic idea of constraining self-attention to fixed or learned windows is likely to become a central topic of research and therefore I tried implementing it. I would like to hear the community's opinions:

1. Does this make sense at all?
2. I am currently working on the 1-D implementation which already beats vanilla and relative Transformer (according to the paper), but the 2-D case seems to bring even more improvement. Should we also implement that? The extension shouldn't be too hard.

3. I added default model names in the `catalog` but I am not sure thats necessary. If we are happy with it we might be able to come up with a better name than `TransformerConv`

# Updates
Also implemented 2D convolution which essentially allows the query of each head to attend to (optionally time-step neighbourhood constrained) keys and values of nearby heads. I am now testing by training and will update with results. The paper is not super-clear on whether the winning implementation included head interaction in all encoder layers or merely the first 3 similar to one convolution. My interpretation was that indeed both operations only impacted the 3 first layers so that's what the `catalog` specification does, but I am not a native speaker of English so I may have misunderstood. 

## Some thoughts on 2D convolution.
While the idea of allowing interactions between heads is promising, formalising it as convolution can be misleading. Convolution is reasonable as a neighbourhood constraint mechanism in the time dimension because we have a reasonable inductive bias: words around the current one are more important to its context than words far away from it. However this does not hold true for the `heads` dimension: `HEAD 1` is in no way *closer* to `HEAD 2` than it is to `HEAD 5`. I still implemented it as a cyclic convolution but it feels like an implementation detail, I would expect it to work exactly the same if we attended to any other set of heads.

### CODE REVIEW

**Comment by guillaumekln** [comment]
1. This looks interesting and I think we can get this in. It would be nice if you can validate it with an actual OpenNMT-tf training.
2. It seems there is quite an improvement in the 2D method over 1D. So yes, if you can implement it that would be great!
3. Adding to the catalog is not required. We could always add it later.

**Comment by steremma** [comment]
Good to know! I started training on 10m parallel data on 8 GPUs and will post final results. After 40k steps (25k effective batch size) BLEU is at 34.49 which is in line with vanilla but this only guarantees there are no "breaking" bugs. I am not sure how to validate that it actually does what we expect it to. 

Have you ever considered attention unit tests that check for actual values rather than shapes? I was thinking something like this (stolen from [tensorflow tutorial](https://www.tensorflow.org/tutorials/text/transformer)):

```
temp_k = tf.constant([[10,0,0],
                      [0,10,0],
                      [0,0,10],
                      [0,0,10]], dtype=tf.float32)  # (4, 3)

temp_v = tf.constant([[   1,0],
                      [  10,0],
                      [ 100,5],
                      [1000,6]], dtype=tf.float32)  # (4, 2)

# This `query` aligns with the second `key`,
# so the second `value` is returned.
temp_q = tf.constant([[0, 10, 0]], dtype=tf.float32)  # (1, 3)

# We expect the second value [10, 0] to be returned.
print(temp_v)
```

**Comment by guillaumekln** [comment]
> Have you ever considered attention unit tests that check for actual values rather than shapes?

Testing values is better but it sometimes difficult to do as layers get more complex.

**Comment by steremma** [comment]
# Update on full training results:

**TL/DR some promising signal but I am not yet 100% certain that its correct. The paper claims 0.4 BLEU increment (28.1 -> 28.5), I observe 0.05 (36.09 -> 36.14).**

Potential reasons:
* Bug in my implementation
* Coincidence. Increments of less than 1 BLEU are generally hard to reproduce.
* Something changed in the reporting of BLEU between ONMT v1.x and v2.x.

## Results

We get **36.09** using the baseline.
``` $ cat vanila.log
INFO:tensorflow:BLEU evaluation score: 35.090000
INFO:tensorflow:BLEU evaluation score: 35.290000
INFO:tensorflow:BLEU evaluation score: 35.680000
INFO:tensorflow:BLEU evaluation score: 35.860000
INFO:tensorflow:BLEU evaluation score: 36.020000
INFO:tensorflow:BLEU evaluation score: 36.090000
INFO:tensorflow:BLEU evaluation score: 36.090000
```

We get **36.14** using 2D conv with the hyper-parameters set as in the paper (first 3 layers convolve in time with window 5 in each side and in heads with 1 in each side).
```
$ cat 2DConv.log
INFO:tensorflow:Evaluation result for step 40000: loss = 1.567502 ; perplexity = 4.794657 ; bleu = 34.503531
INFO:tensorflow:Evaluation result for step 80000: loss = 1.540028 ; perplexity = 4.664719 ; bleu = 35.306190
INFO:tensorflow:Evaluation result for step 120000: loss = 1.534401 ; perplexity = 4.638544 ; bleu = 35.409929
INFO:tensorflow:Evaluation result for step 160000: loss = 1.529972 ; perplexity = 4.618048 ; bleu = 35.582360
INFO:tensorflow:Evaluation result for step 200000: loss = 1.527672 ; perplexity = 4.607437 ; bleu = 35.780820
INFO:tensorflow:Evaluation result for step 240000: loss = 1.523784 ; perplexity = 4.589561 ; bleu = 35.798300
INFO:tensorflow:Evaluation result for step 280000: loss = 1.525808 ; perplexity = 4.598858 ; bleu = 35.884522
INFO:tensorflow:Evaluation result for step 320000: loss = 1.528061 ; perplexity = 4.609231 ; bleu = 35.916077
INFO:tensorflow:Evaluation result for step 360000: loss = 1.527096 ; perplexity = 4.604784 ; bleu = 35.856219
INFO:tensorflow:Evaluation result for step 400000: loss = 1.526644 ; perplexity = 4.602703 ; bleu = 36.050267
INFO:tensorflow:Evaluation result for step 440000: loss = 1.524822 ; perplexity = 4.594327 ; bleu = 36.146093
```

I also tested 1 Conv (so no head interactions) which actually dropped in comparison to baseline at **35.87**:
```
$ cat 1DConv.log
INFO:tensorflow:Evaluation result for step 40000: loss = 1.562896 ; perplexity = 4.772621 ; bleu = 34.433541
INFO:tensorflow:Evaluation result for step 80000: loss = 1.542461 ; perplexity = 4.676084 ; bleu = 35.273885
INFO:tensorflow:Evaluation result for step 120000: loss = 1.532457 ; perplexity = 4.629538 ; bleu = 35.523665
INFO:tensorflow:Evaluation result for step 160000: loss = 1.531114 ; perplexity = 4.623326 ; bleu = 35.648539
INFO:tensorflow:Evaluation result for step 200000: loss = 1.526842 ; perplexity = 4.603616 ; bleu = 35.757203
INFO:tensorflow:Evaluation result for step 240000: loss = 1.527426 ; perplexity = 4.606307 ; bleu = 35.781695
INFO:tensorflow:Evaluation result for step 280000: loss = 1.525511 ; perplexity = 4.597494 ; bleu = 35.877433
```

## Experiment parameters
Everything was based on vanilla Transformer, all default parameters. The dataset only consisted on 8m parallel sentences I found online with no preprocessing. We trained on 8 V-100 GPUs.

## Next steps
It would be great if @guillaumekln could take a very critical look at the code. Ideally we could write a couple of value checked unit tests at least for attention modules at some point.

**Comment by guillaumekln** [comment]
Thanks for these experiments. For how long did you train the baseline model?

**Comment by steremma** [comment]
No problem. All variations were trained for ~ 2 days on the same infrastructure.

**Comment by guillaumekln** [comment]
I took some time to review the implementation and verify it against the paper. It seems correct to me but of course it is easy to miss an important detail.

What was the language pair and data size that you used for the experiment? If you have some free compute time, it would indeed be useful to retrain the baseline with the same version.

**Comment by steremma** [comment]
All experiments used 7.5 million parallel sentences from open sources in the JA -> EN direction. Regarding paper details there is something unclear to me: Does the head interaction (2D conv) also happen only in the first 3 layers like the 1D, or maybe all layers of the encoder? I went with the former in the `catalog` implementation (and my experiment) but I am not sure what the authors meant. 

I can indeed re-run the baseline and report back probably within a week.

**Comment by guillaumekln** [comment]
> Regarding paper details there is something unclear to me: Does the head interaction (2D conv) also happen only in the first 3 layers like the 1D, or maybe all layers of the encoder?

From my understanding they applied both on the 3 lowest encoder layers.

**Comment by steremma** [comment]
On the topic of unit testing, I realised your point upon further thought: it is indeed difficult to check when layers with random initialization come into play. I was wondering, would something like this be of interest?

```
def testExact()
    class MockMultiHeadAttention(transformer.MultiHeadAttention):
        """Eliminate stochasticity to observe results."""
        def __init__(self, num_heads, num_units, **kwargs):
            super(MockMultiHeadAttention, self).__init__(num_heads, num_units, **kwargs)
            self.linear_queries = tf.identity
            self.linear_keys = tf.identity
            self.linear_values = tf.identity
            self.linear_output = tf.identity

    # Single head, 5 units.
    attention = MockMultiHeadAttention(1, 5)

    # Single example in the batch, 4 time steps.
    # Extreme values to guarantee "clean" softmax.
    # With unconstrained attention, t == 2 will be dominated by t == 0.
    queries = tf.constant([[
        [0.0, 500.0, 0.0, 0.0, 0.0],
        [0.0, 0.0, 0.0, 0.0, 300.0],
        [0.0, 200.0, 0.0, 0.0, 0.0],
        [100.0, 0.0, 0.0, 0.0, 0.0]
    ]])
    context, _ = attention(queries, mask=mask)
    context = self.evaluate(context)
    self.assertAllEqual(context[0][2], [0.0, 500.0, 0.0, 0.0, 0.0])

    # This time t == 2 cannot attend to t == 0.
    attention = MockMultiHeadAttention(1, 5, attention_span=1)
    context, _ = attention(queries, mask=None)
    context = self.evaluate(context)
    self.assertAllEqual(context[0][2], [0.0, 200.0, 0.0, 0.0, 0.0])
```

This approach could also be interesting to check other variations of attention (`relative_positions` for example).

**Comment by steremma** [comment]
Running vanilla with the same version (2.5) yields the same results more or less. I am updating the previous comment with the new copy which looks more similar.

**Comment by guillaumekln** [comment]
> On the topic of unit testing, I realised your point upon further thought: it is indeed difficult to check when layers with random initialization come into play. I was wondering, would something like this be of interest?

Usually I try to extract non-parametrized operations into functions and test that instead. For example in your code you should probably test `conv_itself` and the masking operations independently. This is usually where numerical mistakes happen. On the other hand, the most common issue in layers is shape mismatch. So testing that they run without error is usually a good enough test.

> Running vanilla with the same version (2.5) yields the same results more or less. I updated the previous comment outlining results to reflect this. I cannot spend more compute for full experiments but it might be interesting to check another language pair with a standard dataset (WMT) as mine is hand crafted and includes proprietary samples. Ideally the very same one the authors used to report results.

Thanks! So the current conclusion is that this feature has apparently no impact. Do you know if this is implemented elsewhere?

**Comment by steremma** [comment]
> Usually I try to extract non-parametrized operations into functions and test that instead. For example in your code you should probably test `conv_itself` and the masking operations independently. This is usually where numerical mistakes happen. On the other hand, the most common issue in layers is shape mismatch. So testing that they run without error is usually a good enough test.

The testing idea makes sense, I can indeed extract that function and test it separately.

> Thanks! So the current conclusion is that this feature has apparently no impact. Do you know if this is implemented elsewhere?


I agree that the change seems to have no impact on my specific experiment. I am not aware of an implementation as none is mentioned in the paper itself, nor at [paperswithcode](https://paperswithcode.com/paper/convolutional-self-attention-networks) where I normally check these things. I could try pinging the authors.

At the same time, I would not read too much into these results as my dataset is vastly different from the one used in the paper and standard benchmarks. For reference the authors claim an improvement from **28.1 to 28.5** whereas even my baseline is at **36**. I did not care to control about data because the point of the exercise on my end was to check implementation correctness rather than reproducing the paper's result. 

In order to conclude on impact convincingly we could test on a standard dataset if one is available.

**Comment by steremma** [comment]
**Update after discussing with the 1st author of the reference paper (Mr. Baosong Yang):**

1. The implementation is correct (line 297-332 in `layers/transformer.py`).

2. Their work was written on top of a 2 year old [version of TMHMT](https://github.com/THUNLP-MT/THUMT/tree/6611b322295ce9673d650b1d0aa89118fb263f58). It could be that minor improvements or bug fixes in the last couple of years are hiding their observed effect.

3. A few hyper-parameters where different that ONMT default ones both in baseline and their proposed architecture (`dropout, embedding sharing level`).

4. Our datasets are very different both in terms of size and distribution. To quote Mr. Yang:
> Concerning Ja-En translation task, we had ever conducted experiments on a subtitle corpus JESC (link: https://nlp.stanford.edu/projects/jesc/). Experimental results shown that our method has no impact on the performance, and self-attention architecture even yields lower BLEU score than its RNN counterpart.

**Moving forward:** I agree that this work does not seem to have any significant effect on training (no trainable parameters are introduced but also no improvement is observed). However both constraining attention and enabling head interaction can be seen as useful independent modules to experiment with. We could therefore drop the model descriptions from `catalog` and default those parameters to vanilla, but enable the user to experiment with them through the config `yaml`.

Alternatively we drop this PR altogether and consider this exercise a valuable lesson learned. 

@guillaumekln, what do you think?

**Comment by guillaumekln** [comment]
Thanks for the detective work @steremma. With all this information, I'm a bit reluctant to merge this for now.

> However both constraining attention and enabling head interaction can be seen as useful independent modules to experiment with.

This is true. I'm thinking we should add an abstraction layer around the attention computation. For example, having an `Attention` base class that can be extended to `DotProductAttention`, `RelativeAttention`, etc., and pass this class as argument to the top-level `Transformer` model. This would make it easier to add custom attention mechanism without needing to forward the arguments in several places.

**Comment by steremma** [comment]
> With all this information, I'm a bit reluctant to merge this for now.

I understand

> I'm thinking we should add an abstraction layer around the attention computation. For example, having an `Attention` base class that can be extended to `DotProductAttention`, `RelativeAttention`, etc., and pass this class as argument to the top-level `Transformer` model.

Yes that makes a ton of sense. Both features introduced in this PR would fit as arguments to that class since they are both attention related. 


Please let me know if I can in some way help in that effort, and thanks for taking the time to review this work.

**Comment by guillaumekln** [comment]
> This is true. I'm thinking we should add an abstraction layer around the attention computation. For example, having an Attention base class that can be extended to DotProductAttention, RelativeAttention, etc., and pass this class as argument to the top-level Transformer model. This would make it easier to add custom attention mechanism without needing to forward the arguments in several places.

I did not forget about this but it is difficult to provide an interface that is both extensible enough and not overly complex. And it is admittedly not a priority at the moment.

**Comment by naginoasukara** [comment]
Hi, I'm so excited to see your work. Could you tell me how to make a log via opennmt-tf? 
like $ cat vanila.log | grep BLEU
thanks a lot!
